Started handling messages from background isolates.

gaaclarke opened the pull request:
**companion engine pr:** https://github.com/flutter/engine/pull/35174
**design doc:** https://docs.google.com/document/d/1yAFw-6kBefuurXWTur9jdEUAckWiWJVukP1Iay8ehyU/edit#
**issue:** https://github.com/flutter/flutter/issues/13937

## example
### main.dart
```dart
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';
import 'dart:isolate';

void isolateMain(BackgroundIsolateBinding binding) async {
  BackgroundIsolateBinding.initializeBackgroundIsolate(binding);
  const BasicMessageChannel channel =
      BasicMessageChannel('dev.flutter.aaclarke.echo', StandardMessageCodec());
  final String result = await channel.send('hello');
  print(result);
}

void main() {
  final BackgroundIsolateBinding binding =
      BackgroundIsolateBinding.initializeRootIsolate();
  Isolate.spawn(isolateMain, binding);
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(title: 'Isolate Platform Channels'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({super.key, required this.title});
  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Container(),
    );
  }
}
```

### AppDelegate.swift
```swift
import UIKit
import Flutter
import os.log

@UIApplicationMain
@objc class AppDelegate: FlutterAppDelegate {
  override func application(
    _ application: UIApplication,
    didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?
  ) -> Bool {
    let registrar = self.registrar(forPlugin: "echo");
    let channel = FlutterBasicMessageChannel(name: "dev.flutter.aaclarke.echo", binaryMessenger: registrar!.messenger());
    channel.setMessageHandler { (input, reply) in
      if let input_string = input as? String? {
        os_log("%@", input_string ?? "null");
      } else {
        os_log("unrecognized message");
      }
      reply(input)
    }
    GeneratedPluginRegistrant.register(with: self)
    return super.application(application, didFinishLaunchingWithOptions: launchOptions)
  }
}
```

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

gaaclarke committed:
Started handling messages from background isolates.

Hixie commented:
The overall idea seems fine. I'm not sure I understand the specifics of the implementation on the Dart side. Maybe we can go over the PR together on VC?

Hixie commented:
@gaaclarke here's a proposed reworking of the API as discussed in our chat:
https://github.com/flutter/flutter/compare/master...Hixie:flutter:isolate-platform-channels

It's semantically equivalent to the original code, just presented differently.

gaaclarke committed:
- moved to `get rootIsolateId`
- introduced RootIsolateToken
- Added apropos error when not having registered the background isolate
- Switched to BackgroundIsolateBinaryMessenger
- Added getter for the rootIsolateToken to ServicesBinding

gaaclarke commented:
> @gaaclarke here's a proposed reworking of the API as discussed in our chat: [master...Hixie:flutter:isolate-platform-channels](https://github.com/flutter/flutter/compare/master...Hixie:flutter:isolate-platform-channels)
> 
> It's semantically equivalent to the original code, just presented differently.

Thanks, I've incorporated most of it now to see how it worked out.  I still need to clean it up and incorporate your docs.  I made 2 changes:
1) Not requiring the binary messenger in the channel constructor as we discussed in https://github.com/flutter/flutter/commit/04194fd1ada2a94125b16fc2a9d21e2eed715ebc#commitcomment-81568761
1) I put the logic for acquiring the the RootIsolateToken on that class and started exporting it from `services/platform_channels.dart`.  I wanted to make sure that importing services was sufficient to allowing one to use platform channels on background isolates.

Hixie commented on packages/flutter/lib/src/services/platform_channel.dart:
```diff
@@ -123,6 +126,79 @@ void _debugRecordDownStream(String channelTypeName, String name,
   _debugLaunchProfilePlatformChannels();
 }
 
+class BackgroundIsolateBinaryMessenger extends BinaryMessenger {
```
this class should have a private constructor so that people can't make instances that don't match the `instance`

gaaclarke replied:
done

Hixie commented on packages/flutter/lib/src/services/platform_channel.dart:
```diff
@@ -123,6 +126,79 @@ void _debugRecordDownStream(String channelTypeName, String name,
+  final Map<int, Completer<ByteData?>> _completers = <int, Completer<ByteData?>>{};
+  int _messageCount = 0;
+
+  static BinaryMessenger? _backgroundIsolateBinaryMessenger;
```
for consistency with other cases of this pattern we should call this _instance

gaaclarke replied:
done

Hixie commented on packages/flutter/lib/src/services/platform_channel.dart:
```diff
@@ -123,6 +126,79 @@ void _debugRecordDownStream(String channelTypeName, String name,
+
+  static BinaryMessenger? _backgroundIsolateBinaryMessenger;
+
+  static void ensureInitialized(ui.RootIsolateToken token) {
```
this will need api docs (you can copy them from my branch)

gaaclarke replied:
done

Hixie commented on packages/flutter/lib/src/services/platform_channel.dart:
```diff
@@ -123,6 +126,79 @@ void _debugRecordDownStream(String channelTypeName, String name,
+      ui.PlatformDispatcher.instance.registerBackgroundIsolate(token);
+      final BackgroundIsolateBinaryMessenger portBinaryMessenger = BackgroundIsolateBinaryMessenger();
+      _backgroundIsolateBinaryMessenger = portBinaryMessenger;
+      portBinaryMessenger.receivePort.listen((dynamic message) {
```
i would advise using the pattern i had on my branch to make this a static method rather than a closure. That'll avoid accidentally closing over things we don't want to close over, but also generally the code will be easier to reason about (and it makes the satisfying pattern of having both the send and the receive logic be parallel).

Hixie commented on packages/flutter/lib/src/services/platform_channel.dart:
```diff
@@ -123,6 +126,79 @@ void _debugRecordDownStream(String channelTypeName, String name,
+    }
+  }
+
+  static BinaryMessenger? get instance => _backgroundIsolateBinaryMessenger;
```
based on experience with other `instance` getters, we should make this non-nullable (nobody is going to null-check it anyway, they'll just use `!`) and then have it throw if it's null (my branch has an easy way to do that, but you may also want to consider a more elaborate solution in the vein of `BaseBinding.instance`).

Hixie replied:
(also you'll need dartdocs, feel free to copy the ones from my branch)

Hixie replied:
(also per https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#order-other-class-members-in-a-way-that-makes-sense and generally to match the conventions used elsewhere, we should probably use the order 1) getter, 2) private field, 3) ensureInitialized)

gaaclarke replied:
done

Hixie commented on packages/flutter/lib/src/services/platform_channel.dart:
```diff
@@ -123,6 +126,79 @@ void _debugRecordDownStream(String channelTypeName, String name,
+
+  @override
+  Future<void> handlePlatformMessage(String channel, ByteData? data, ui.PlatformMessageResponseCallback? callback) {
+    throw UnimplementedError();
```
suggest passing an argument to this, see my branch for one option

gaaclarke replied:
done

Hixie commented on packages/flutter/lib/src/services/platform_channel.dart:
```diff
@@ -123,6 +126,79 @@ void _debugRecordDownStream(String channelTypeName, String name,
+
+  @override
+  void setMessageHandler(String channel, MessageHandler? handler) {
+    throw UnimplementedError();
```
suggest using UnsupportedError instead, with an argument, again feel free to copy from my branch

gaaclarke replied:
done

Hixie commented on packages/flutter/lib/src/services/platform_channel.dart:
```diff
@@ -123,6 +126,79 @@ void _debugRecordDownStream(String channelTypeName, String name,
+  if (ui.RootIsolateToken.instance == null) {
+    assert(() {
+      if (BackgroundIsolateBinaryMessenger.instance == null) {
+        throw Exception(
```
in general we should use specific subclasses of Exception in framework code. Also this should be an `Error` not an `Exception`. (IIRC in my branch I used ArgumentError.)

gaaclarke replied:
done, StateError now

Hixie commented on packages/flutter/lib/src/services/platform_channel.dart:
```diff
@@ -123,6 +126,79 @@ void _debugRecordDownStream(String channelTypeName, String name,
+    assert(() {
+      if (BackgroundIsolateBinaryMessenger.instance == null) {
+        throw Exception(
+            'Must call BackgroundIsolateBinaryMessenger.ensureInitialized.');
```
If possible I recommend providing more helpful error messages.

Remember the mantra, every error message is an opportunity to make a developer love Flutter.

gaaclarke replied:
Sorry, missed this.  Done.  I didn't know what other information to provide but I made it more clearly written.

Hixie commented on packages/flutter/lib/src/services/platform_channel.dart:
```diff
@@ -123,6 +126,79 @@ void _debugRecordDownStream(String channelTypeName, String name,
+      return true;
+    }());
+    return BackgroundIsolateBinaryMessenger.instance!;
+  } else {
```
(trivial nit: avoid else after return)

gaaclarke replied:
done

Hixie commented on packages/flutter/lib/src/services/platform_channel.dart:
```diff
@@ -123,6 +126,79 @@ void _debugRecordDownStream(String channelTypeName, String name,
+    final int messageIdentifier = ++_messageCount;
+    _completers[messageIdentifier] = completer;
+    ui.PlatformDispatcher.instance.sendPortPlatformMessage(
+        channel, message, messageIdentifier, receivePort.sendPort);
```
trivial nit: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#if-you-have-a-newline-after-some-opening-punctuation-match-it-on-the-closing-punctuation

gaaclarke replied:
done

Hixie commented on packages/flutter/lib/src/services/platform_channel.dart:
```diff
@@ -123,6 +126,79 @@ void _debugRecordDownStream(String channelTypeName, String name,
+  @override
+  Future<ByteData?>? send(String channel, ByteData? message) {
+    final Completer<ByteData?> completer = Completer<ByteData?>();
+    final int messageIdentifier = ++_messageCount;
```
trivial nit: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#prefer--over-

gaaclarke replied:
done

Hixie commented on packages/flutter/lib/src/services/platform_channel.dart:
```diff
@@ -123,6 +126,79 @@ void _debugRecordDownStream(String channelTypeName, String name,
+          final int identifier = args[0] as int;
+          final Uint8List bytes = args[1] as Uint8List;
+          final ByteData byteData = ByteData.sublistView(bytes);
+          final Completer<ByteData?> completer =
+              portBinaryMessenger._completers[identifier]!;
+          portBinaryMessenger._completers.remove(identifier);
+          completer.complete(byteData);
```
You can simplify this to:
```suggestion
          _completers.remove(identifier)!.complete(byteData);
```

gaaclarke replied:
done

Hixie commented on packages/flutter/lib/src/services/platform_channel.dart:
```diff
@@ -123,6 +126,79 @@ void _debugRecordDownStream(String channelTypeName, String name,
 }
 
+class BackgroundIsolateBinaryMessenger extends BinaryMessenger {
+  final ReceivePort receivePort = ReceivePort();
```
can we make this private? I'm a bit wary of having people fiddle with it from outside the class...

gaaclarke replied:
done

Hixie commented on packages/flutter/lib/src/services/platform_channel.dart:
```diff
@@ -162,8 +238,7 @@ class BasicMessageChannel<T> {
 
   /// The messenger which sends the bytes for this channel, not null.
   BinaryMessenger get binaryMessenger {
```
we should also update the API docs for all these classes to mention that they work on both the root isolate and background isolates, and what method to call to make sure they will work on the root isolate

Hixie replied:
(my branch has an example of what that could look like but it'll need to be tweaked to the new semantics we discussed)

gaaclarke replied:
Sorry, missed this, done.

Hixie commented:
> Thanks, I've incorporated most of it now to see how it worked out. I still need to clean it up and incorporate your docs.

Oops, I missed that you hadn't finished yet, feel free to ignore my review comments until you have. :-D

Hixie commented on packages/flutter/lib/src/services/platform_channel.dart:
```diff
@@ -123,6 +126,79 @@ void _debugRecordDownStream(String channelTypeName, String name,
+}
+
+BinaryMessenger _findBinaryMessenger() {
+  if (ui.RootIsolateToken.instance == null) {
```
I guess this really doesn't matter because nobody but us is ever going to use this API at the end of the day, but if this was a more widely-used API I would definitely suggest having separate "isRootIsolate" vs "RootIsolateToken.instance" affordances, because the former is much clearer than comparing the latter to null.

gaaclarke replied:
I tried `isRootIsolate` but Dan had some misgivings about talking about "root isolates" in the context of the web.  So now it's "get useBackgroundIsolateBinaryMessenger".

Hixie commented on dev/integration_tests/channels/lib/src/basic_messaging.dart:
```diff
@@ -78,6 +79,37 @@ Future<TestStepResult> basicStandardHandshake(dynamic message) async {
+
+Future<TestStepResult> basicBackgroundStandardEcho(Object message) async {
+  final ReceivePort receivePort = ReceivePort();
+  Isolate.spawn(_basicBackgroundStandardEchoMain, <Object>[RootIsolateToken.instance!, message, receivePort.sendPort]);
```
FWIW, as a general rule, in the root isolate we should never access dart:ui directly, only via the bindings. That's why in my branch I indirected via the binding to get the RootIsolateToken. Otherwise we have no way to mock dart:ui in the tests when we need to. (That's also why I had RootIsolateToken be a "plain old data" class with the logic to get one on the PlatformDispatcher; the "plain old data" classes can be used directly in tests, and the PlatformDispatcher is mockable, but class statics cannot and are not.)

gaaclarke replied:
Makes sense why you'd put it on the `ServiceBinding`, did it.

I don't know what you mean by "plain old data".  Your patch didn't have the changes to the engine so it's hard to see what you mean.

gaaclarke replied:
I had to introduce "useBackgroundIsolateBinaryMessenger" to make the existing mocks work.  Since we have test coverage with integration tests I'm not sure if we'll ever have to mock the GetRootIsolateToken.

dnfield commented on packages/flutter/lib/src/services/binding.dart:
```diff
@@ -82,6 +82,14 @@ mixin ServicesBinding on BindingBase, SchedulerBinding {
+  /// This property is primarily intended for use with
+  /// [BackgroundIsolateBinaryMessenger.ensureInitialized], which takes a
+  /// [RootIsolateToken] as its argument.
+  ui.RootIsolateToken? get rootIsolateToken => ui.RootIsolateToken.instance;
```
Consider explaining that this will return null if you're not currently on the root isolate.

Also, consider _not_ exposing this at all in the framework's API. Maybe a nicer name for this would be something like `ui.RootIsolateData.token` and `bool get ui.RootIsolteData.isRootIsolate => token != null` or something.

gaaclarke replied:
I've reworked this and added a `isRootIsolate` to fix mocks for tests.  Everything in `dart:ui` is technically available to users of the framework, so I can't really hide it.  I'm still chasing down making all the web tests working and mocked bindings functions correctly.  Once I have all the lights green I can come back and make sure we are happy with it.

gaaclarke committed:
added docstrings

gaaclarke committed:
Fixed web and tests detection of the binary messenger

gaaclarke committed:
turned off the android tests

gaaclarke commented:
@dnfield I got everything cleaned up and all the tests passing and ready to review.  I introduced `bool get useBackgroundIsolateBinaryMessenger` to handle the web and tests case.

gaaclarke committed:
ian doc and error message feedback

gaaclarke commented:
> It looks like this pull request may not have tests. Please make sure to add tests before merging. If you need [an exemption](https://github.com/flutter/flutter/wiki/Tree-hygiene#tests) to this rule, contact Hixie on the #hackers channel in [Chat](https://github.com/flutter/flutter/wiki/Chat) (don't just cc him here, he won't see it! _He's on Discord!_).

False positive, there are integration tests.

dnfield reviewed (approved):
LGTM

gaaclarke commented:
Over the course of responding to feedback I broke the integration test which apparently doesn't run as a presubmit.  Reverting to open the tree.
